v8go.cc

```diff
@@ -148,79 +165,152 @@ IsolateHStatistics IsolationGetHeapStatistics(IsolatePtr ptr) {
+  }
+
+  ValuePtr goFunctionCallback(int ctxref, int cbref, const ValuePtr* args,
+                              int args_count);
+  ValuePtr val_ptr =
+      goFunctionCallback(ctx_ref, callback_ref, args, args_count);
+  if (val_ptr != nullptr) {
+    m_value* val = static_cast<m_value*>(val_ptr);
+    info.GetReturnValue().Set(val->ptr.Get(iso));
+  } else {
+    info.GetReturnValue().SetUndefined();
+  }
 }
 
-void ObjectTemplateSetValue(ObjectTemplatePtr ptr,
-                       const char* name,
-                       ValuePtr val_ptr,
-                       int attributes) {
-  LOCAL_OBJECT_TEMPLATE(ptr);
+TemplatePtr NewFunctionTemplate(IsolatePtr iso_ptr, int callback_ref) {
+  Isolate* iso = static_cast<Isolate*>(iso_ptr);
+  Locker locker(iso);
+  Isolate::Scope isolate_scope(iso);
+  HandleScope handle_scope(iso);
 
-  Local<String> prop_name =
-      String::NewFromUtf8(iso, name, NewStringType::kNormal).ToLocalChecked();
-  m_value* val = static_cast<m_value*>(val_ptr);
-  object_template->Set(prop_name, val->ptr.Get(iso), (PropertyAttribute)attributes);
-}
+  m_ctx* ctx = static_cast<m_ctx*>(iso->GetData(0));
+  Local<Context> local_ctx = ctx->ptr.Get(iso);
+  Context::Scope context_scope(local_ctx);
 
-void ObjectTemplateSetObjectTemplate(ObjectTemplatePtr ptr, const char* name, ObjectTemplatePtr obj_ptr, int attributes) {
-  LOCAL_OBJECT_TEMPLATE(ptr);
+  Local<Array> cbData = Array::New(iso, 1);
+  Maybe<bool> set = cbData->Set(local_ctx, 0, Integer::New(iso, callback_ref));
 
-  Local<String> prop_name =
-      String::NewFromUtf8(iso, name, NewStringType::kNormal).ToLocalChecked();
-  m_object_template* obj = static_cast<m_object_template*>(obj_ptr);
-  object_template->Set(prop_name, obj->ptr.Get(iso), (PropertyAttribute)attributes);
+  m_template* ot = new m_template;
+  ot->iso = iso;
+  ot->ptr.Reset(iso,
+                FunctionTemplate::New(iso, FunctionTemplateCallback, cbData));
+  return static_cast<TemplatePtr>(ot);
 }
 
 /********** Context **********/
 
 ContextPtr NewContext(IsolatePtr iso_ptr,
-                      ObjectTemplatePtr global_template_ptr) {
+                      TemplatePtr global_template_ptr,
+                      int ref) {
   Isolate* iso = static_cast<Isolate*>(iso_ptr);
   Locker locker(iso);
   Isolate::Scope isolate_scope(iso);
   HandleScope handle_scope(iso);
 
   Local<ObjectTemplate> global_template;
   if (global_template_ptr != nullptr) {
-    m_object_template* ob =
-        static_cast<m_object_template*>(global_template_ptr);
-    global_template = ob->ptr.Get(iso);
+    m_template* ob = static_cast<m_template*>(global_template_ptr);
+    global_template = ob->ptr.Get(iso).As<ObjectTemplate>();
   } else {
     global_template = ObjectTemplate::New(iso);
   }
 
-  iso->SetCaptureStackTraceForUncaughtExceptions(true);
+  Local<Context> local_ctx = Context::New(iso, nullptr, global_template);
+  local_ctx->SetEmbedderData(1, Integer::New(iso, ref));
```

Comment:
I had to dig a bit to understand how this was used -- could you add some comments in here?